### PR TITLE
Ajt/util comp/shared components

### PR DESCRIPTION
**Pull request by Jordan-Trimarchi**
Adds file for shared components including 2 functions. One to render a 5 star display filled to the nearest quarter. The other takes in an object containing counts of various review ratings and averages it into an average rating number between 1-5.

Adds a file for the ratings breakdown element of the R&R widget.

changes some link rules per alex's suggestion.

minor formatting changes to please the airbnb gods

### CODE REVIEW

File: client/src/components/SharedComponents.jsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  return (
+    <>
+      {starFill.map((percent) => <span key={Math.random()} className={`star star-${percent} fa fa-star`} />)}
+    </>
+  );
+};
+const AverageRatings = (ratings) => {
+  let total = 0;
+  let length = 0;
+  for (const key in ratings) {
```
Comment (adbarc92) [inline client/src/components/SharedComponents.jsx]: We'll probably want to stay away from for-in, as there can be unintended consequences. Basically, it iterates over the whole object, including the prototype, which can cause problems. Object.keys is usually a better solution. It, and some other options, are detailed here: https://www.30secondsofcode.org/articles/s/eslint-refactor-for-in

File: client/src/components/SharedComponents.jsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    starFill[i] = 100;
+  }
+  starFill[starFill.indexOf(0)] = remainingFraction;
+  return (
+    <>
+      {starFill.map((percent) => <span key={Math.random()} className={`star star-${percent} fa fa-star`} />)}
+    </>
+  );
+};
+const AverageRatings = (ratings) => {
```
Comment (adbarc92) [inline client/src/components/SharedComponents.jsx]: This should probably be a utility function

File: client/src/components/SharedComponents.jsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  let length = 0;
+  for (const key in ratings) {
+    total += key * ratings[key];
+    length += Number(ratings[key]);
+  }
+  return (total / length).toFixed(2);
+};
+
+export default StarRating;
+export { AverageRatings };
```
Comment (adbarc92) [inline client/src/components/SharedComponents.jsx]: Adding propTypes would include
`StarRating.propTypes = {
rating: PropTypes.number.isRequired
}`

File: client/src/components/R&R/ReviewsBreakdown.jsx [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import React, { useState, useContext } from 'react';
+import StarRating, { AverageRatings } from '../SharedComponents';
+import { ProductContext } from '../../context';
+
+const ReviewsBreakdown = () => {
+  const { reviews, reviewsMeta, reviewSort, setReviewSort } = useContext(ProductContext);
+  return (
+    <>
+      <div>ratings</div>
```
Comment (adbarc92) [inline client/src/components/R&R/ReviewsBreakdown.jsx]: 10/10 medical oncologists agree that capitalization inhibits cancer cell growth

File: client/src/__tests__/utils.test.js [hunk 5]
```diff
@@ -22,4 +22,4 @@ describe('composeParams', () => {
     const expected = '?product_id=1813';
     expect(result).toBe(expected);
   });
-});
\ No newline at end of file
+});
```
Comment (adbarc92) [inline client/src/__tests__/utils.test.js:25]: Was this just broken and missing a set of closing parens/brackets?

File: client/src/components/SharedComponents.jsx (hunk 3 above)
Comment (adbarc92) [reply client/src/components/SharedComponents.jsx]: On the other hand, I am not entirely committed to PropTypes. We should discuss tomorrow whether to go with it or disable the ESLint rule.

**Review by adbarc92 — CHANGES_REQUESTED** [review]
The `For In` Loop should probably be changed, and we should figure out prop types. Otherwise, looks good. I'd congratulate you, but that might establish a precedent for kindness, and I'm allergic to that.

File: client/src/__tests__/utils.test.js (hunk 5 above)
Comment (Jordan-Trimarchi) [reply client/src/__tests__/utils.test.js:25]: this was probably one of the minor changes i made to appease the linter. like removing a blank line or something like that.

File: client/src/components/R&R/ReviewsBreakdown.jsx [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import React, { useState, useContext } from 'react';
+import StarRating from '../SharedComponents';
+import { averageRatings } from '../../utils';
+import { ProductContext } from '../../context';
+
+const ReviewsBreakdown = () => {
+  const { reviews, reviewsMeta, reviewSort, setReviewSort } = useContext(ProductContext);
+  console.log(reviewsMeta.ratings);
```
Comment (adbarc92) [inline client/src/components/R&R/ReviewsBreakdown.jsx]: Remove console.logs

**Review by adbarc92 — COMMENTED** [review]
Things remain

**Review by adbarc92 — APPROVED** [review]
Looks good

**Pull request merged**